cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -233,6 +235,282 @@ struct DivideChecked {
   }
 };
 
+struct Power {
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_signed_integer<T> Call(KernelContext* ctx, Arg0 left, Arg1 right) {
+    Arg0 result = 1;
+    if (left == 0) {
+      if (right < 0) {
+        ARROW_PREDICT_FALSE(MultiplyWithOverflow(2, std::numeric_limits<Arg0>::max(), &result));
+        ctx->SetStatus(Status::Invalid("divide by zero"));
+        return result;
+      } else if (right == 0) {
+        return 1;
+      }
+      return 0;
+    } else if (right == INFINITY && abs(left) > 1) {
+      ARROW_PREDICT_FALSE(MultiplyWithOverflow(2, std::numeric_limits<Arg0>::max(), &result));
+      return result;
+    } else if (right == -INFINITY && abs(left) > 1) {
+      return 0;
+    }
+
+    if (right > 0) {
+      for (Arg1 i = 0; i < right; i++) {
+        if (ARROW_PREDICT_FALSE(MultiplyWithOverflow(result, left, &result))) {
+          ctx->SetStatus(Status::Invalid("overflow"));
+        }
+      }
+    } else {
+      for (Arg1 i = 0; i < -right; i++) {
```

Comment:
Again, negative exponent is illegal for integer power. This function is returning T, which is an integer.
We'd better reference implementations from other software first, e.g., https://numpy.org/doc/stable/reference/generated/numpy.power.html

Reply:
I'm not sure about the negative integer exponents. Feels mathematically wrong :). I'll think about it a bit.

Reply:
R supports power to negative integers so I'd propose to have negative integer exponents enabled.

Reply:
I *think* the reason this is not allowed in numpy is because it would not be type stable (an int to a negative power would give a float, but an int to a positive power preserves the integer dtype; so allowing both would give value-dependent result type). 

For example plain Python also allows it:

```
In [16]: 2 ** 2
Out[16]: 4

In [17]: 2 ** -2
Out[17]: 0.25
```

but for plain Python there is less need to type stability. 

I think for Arrow we want to have kernels with defined input / output types?

Reply:
Good point @jorisvandenbossche.
So from users perspective negative exponents will return integer rounded results which is not bad if it doesn't surprise the user. But the user might want to go from integers to floats without converting the input first.
Shall we keep the current kernels and add an option to return floats from ints when requested? Or should it be an opt out?

Reply:
It will be easy if we support an option to set explicitly the return value type, like numpy `dtype` .
Otherwise, we have to scan the whole exponent array to see if there're negative exponents before deciding the output type.

Reply:
> So from users perspective negative exponents will return integer rounded results which is not bad if it doesn't surprise the user.

I think this will be quite surprising. I would rather raise an error.

Reply:
I've switched to raising an error for negative integers.
I'm not sure what to do about type. Should we do type promotion or not?

Reply:
Promotion looks reasonable for power operation. Always return `int64` for signed integers, `uint64` for unsigned.